smarts/env/wrappers/metrics.py

```diff
@@ -0,0 +1,273 @@
+                        counts=Counts(),
+                        costs=Costs(),
+                    ),
+                    cost_funcs=CostFuncs(),
+                )
+                for agent_name in self._cur_agents
+            }
+
+        return obs
+
+    def records(self) -> Dict[str, Dict[str, Record]]:
+        """
+        Fine grained performance metric for each agent in each scenario.
+
+        Returns:
+            Dict[str, Dict[str, Record]]: Performance record in a nested
+                dictionary for each agent in each scenario.
+
+        Example::
+
+        >> records()
+        >> {
+                scen1: {
+                    agent1: Record(counts, costs),
+                    agent2: Record(counts, costs),
+                },
+                scen2: {
+                    agent1: Record(counts, costs),
+                },
+            }
+        """
+
+        records = {}
+        for scen, agents in self._records.items():
+            records[scen] = {}
+            for agent, data in agents.items():
+                records[scen][agent] = copy.deepcopy(data.record)
+
+        return records
+
+    def score(self) -> Dict[str, float]:
+        """
+        An overall performance score achieved on the wrapped environment.
+        """
+        counts_list, costs_list = zip(
+            *[
+                (data.record.counts, data.record.costs)
+                for agents in self._records.values()
+                for data in agents.values()
+            ]
+        )
+        counts_tot = functools.reduce(lambda a, b: _add_dataclass(a, b), counts_list)
+        costs_tot = functools.reduce(lambda a, b: _add_dataclass(a, b), costs_list)
+
+        _score: Dict[str, float] = {}
+        _score["completion"] = _completion(counts=counts_tot)
+        _score["humanness"] = _humanness(counts=counts_tot, costs=costs_tot)
+        _score["rules"] = _rules(counts=counts_tot, costs=costs_tot)
+        _score["time"] = _time(counts=counts_tot, costs=costs_tot)
+
+        return _score
+
+
+def _check_env(env):
+    def check_intrfc(agent_intrfc):
+        intrfc = {
+            "accelerometer": bool(agent_intrfc.accelerometer),
+            "max_episode_steps": bool(agent_intrfc.max_episode_steps),
+            "neighborhood_vehicles": bool(agent_intrfc.neighborhood_vehicles),
+            "road_waypoints": bool(agent_intrfc.road_waypoints),
+            "waypoints": bool(agent_intrfc.waypoints),
+        }
+        return intrfc
+
+    for agent_name, agent_spec in env.agent_specs.items():
+        intrfc = check_intrfc(agent_spec.interface)
+        if not all(intrfc.values()):
+            raise AttributeError(
+                "Enable {0}'s disabled interface to "
+                "compute its metrics. Current interface is "
+                "{1}.".format(agent_name, intrfc)
+            )
+
+
+T = TypeVar("T", Costs, Counts)
+
+
+def _add_dataclass(first: T, second: T) -> T:
+    new = {}
```

Comment:
A reasonable check would be to test that the instances are of the same class.
```suggestion
def _add_dataclass(first: T, second: T) -> T:
    assert type(first) is type(second)
    new = {}
```

Reply:
Added type assertion.